expand rapids-cmake cpm docs (#613)
rapidsai/rapids-cmake: docs/cpm.rst

CPM Reference
#############

The `rapids_cpm` functions allow projects to find or build dependencies with built-in
tracking of these dependencies for correct export support.

> rapids-cmake package defaults

Reviewer: This sequence is a bit odd. I would prefer
1. rapids-cmake provides packages
2. rapids-cmake encodes information about package versions in a json file; mention that it includes some set of defaults (but don't list here) then go into the format itself
3. overrides
4. pinning and reproducible builds
5. list of defaults

Author: The primary reader I had in mind for this document was a consumer of rapids-cmake and not a developer. So I wanted to first cover what cpm packages we offer, and the features ( overrides, pins ) that interact with the versions.json.

Therefore in those cases the exact details on the versions file are last, since that is the most technical detail that isn't needed unless you are writing your own override

Reviewer: When you say developers, who do you mean? I assume that all consumers of rapids-cmake are developers of other libraries, no?

But assuming I'm misunderstanding the audience you intend, I agree that the sequence seems to make sense for someone who doesn't need to know about overrides.

Author: >  I agree that the sequence seems to make sense for someone who doesn't need to know about overrides.

:+1: 

> When you say developers, who do you mean? I assume that all consumers of rapids-cmake are developers of other libraries, no?

Downstream build-system developers was whom I was thinking about as 'developers'.

When an override for a project exists no :cmake:module:`find_package()<cmake:command:find_package>` search for that project will occur. This is done to make sure that the requested modified version is used.

If a project is listed in multiple override files, the first file values will be used,
and all later calls for that packaged will be ignored.  This "first to record, wins"
approach is used to match FetchContent, and allows parent projects to override child
projects.

If the override is of an existing default project, it only needs to specify the
fields it wants to override.

For example if you wanted to change the version of `fmt` you would do the following:

.. literalinclude:: /packages/example_git_tag_override.json
  :language: json

If you want to specify a completely new project, you need to specify at a minimum all
the required fields:

.. literalinclude:: /packages/example_new_project_override.json
  :language: json

reproducible rapids-cmake builds
################################

The rapids-cmake default `versions.json` uses branch names or git tag names
for dependencies. This is done so that projects can 'live at head' of dependencies.

> This directly conflicts with the concept of reproducible release packages that each
> time they are built contain the exact same byteconde.

Reviewer: There's a typo in bytecode you could fix instead if you prefer the original wording, but I think bytecode is a bit confusing to mention here since traditionally bytecode makes me think of interpreter IR (e.g. JVM or Python interpreter instructions) rather than machine code.
```suggestion
This directly conflicts with the concept of reproducible release packages that each
time they are built produce bitwise-identical artifacts.
```

> :cmake:command:`rapids_cpm_generate_pinned_versions` can be used to generate a set of
> rapids-cmake dependencies from the existing branch names to explicit pinned git SHA1 values. This results in a fully reproducible set of dependencies when building.

Reviewer: ```suggestion
:cmake:command:`rapids_cpm_generate_pinned_versions` can be used to generate a set of explicit
pinned git SHA1 values corresponding to the set of rapids-cmake dependencies in use. This results in a fully reproducible set of dependencies when building.
```

To utilize this behavior in your release CI/CD something like the following needs

> to be setup:

Reviewer: ```suggestion
to be set up:
```

>   1. Enable generation of a pinned versions during builds via

Reviewer: ```suggestion
  1. Enable generation of a pinned versions file during builds via
```

  :cmake:command:`rapids_cpm_generate_pinned_versions` or by specifying the :cmake:variable:`RAPIDS_CMAKE_CPM_PINNED_VERSIONS_FILE`.

>   2. If build is good, create the release branch and commit the generated pinned `versions.json` to the repository

Reviewer: ```suggestion
  2. If the build is good, create the release branch and commit the generated pinned `versions.json` to the repository
```

>   3. Rebuilds of the project using the pinned version are now possible by using the :cmake:variable:`RAPIDS_CMAKE_CPM_OVERRIDE_VERSION_FILE` set to the path of the generated pinned versions file.

Reviewer: ```suggestion
  3. Rebuilds of the project using the pinned version are now possible by setting the :cmake:variable:`RAPIDS_CMAKE_CPM_OVERRIDE_VERSION_FILE` to the path of the generated pinned versions file.
```

> rapids-cmake cpm command line controls

Reviewer: ```suggestion
rapids-cpm command line controls
```

######################################

> rapids-cmake cpm offers multiple command line options to control behavior

Reviewer: ```suggestion
rapids-cpm offers multiple command line options to control behavior.
```

If you need to explicitly state a package must be downloaded and not searched
for locally you enable the variable :cmake:variable:`CPM_DOWNLOAD_<package_name>`.

```

> cmake -DCPM_DOWNLOAD_<package_name>=ON ....

Reviewer: I would suggest a separate subheading that documents the most important underlying CPM options to delineate which options are ours and which come from CPM. Ideally we would link to relevant sections in CPM docs, but AFAIK the only docs are the CPM README and it's not that easy to link to sections, so it may just have to a be a single link to the file saying 'all CPM options can be found here'.

Author: Good idea. I have re-organized the section and would like a re-review.

```

> Some build locations are fully offline and the default packages and override urls

Reviewer: ```suggestion
Some builds are performed fully offline and the default package and override urls
```

can't be used. In those cases you can use the variable :cmake:variable:`RAPIDS_CMAKE_CPM_OVERRIDE_VERSION_FILE` to provide a new versions.json that will be
used instead of ALL overrides specified in that project. This would allow you to

> specifify custom internal urls for all dependencies without modifying the project source code.

Reviewer: ```suggestion
specify custom internal urls for all dependencies without modifying the project source code.
```

```
cmake -DRAPIDS_CMAKE_CPM_OVERRIDE_VERSION_FILE=<abs/path/to/custom/override_versions.json> ....
```

To request the generation of a pinned package override file without having to modify
the project use the :cmake:variable:`RAPIDS_CMAKE_CPM_PINNED_VERSIONS_FILE` variable:

```
cmake -DRAPIDS_CMAKE_CPM_PINNED_VERSIONS_FILE
```

> RAPIDS_CMAKE_CPM_PINNED_VERSIONS_FILE
> 

Reviewer: ```suggestion
```

rapids-cmake package version format
###################################

rapids-cmake uses a JSON file to encode the version of a project and how to download the project.

The JSON format is a root object that contains the ``packages`` object.

The ``packages`` object contains a key/value map of all supported

> packages where the key is the case sensitive name of the project and

Reviewer: ```suggestion
packages where the key is the case-sensitive name of the project and
```

When an override for a project exists no :cmake:module:`find_package()<cmake:command:find_package>` search for that project will occur. This is done to make sure that the requested modified version is used.

If a project is listed in multiple override files, the first file values will be used,
and all later calls for that packaged will be ignored.  This "first to record, wins"
approach is used to match FetchContent, and allows parent projects to override child
projects.

If the override is of an existing default project, it only needs to specify the
fields it wants to override.

For example if you wanted to change the version of `fmt` you would do the following:

.. literalinclude:: /packages/example_git_tag_override.json
  :language: json

If you want to specify a completely new project, you need to specify at a minimum all
the required fields:

.. literalinclude:: /packages/example_new_project_override.json
  :language: json

> reproducible rapids-cmake builds

Reviewer: ```suggestion
Reproducible rapids-cmake builds
```

The rapids-cmake default `versions.json` uses branch names or git tag names
for dependencies. This is done so that projects can 'live at head' of dependencies.
This directly conflicts with the concept of reproducible release packages that each
time they are built produce bitwise-identical artifacts.

:cmake:command:`rapids_cpm_generate_pinned_versions` can be used to generate a set of explicit
pinned git SHA1 values corresponding to the set of rapids-cmake dependencies in use. This results in a fully reproducible set of dependencies when building.

To utilize this behavior in your release CI/CD something like the following needs
to be set up:

  1. Enable generation of a pinned versions file during builds via
  :cmake:command:`rapids_cpm_generate_pinned_versions` or by specifying the :cmake:variable:`RAPIDS_CMAKE_CPM_PINNED_VERSIONS_FILE`.
  2. If the build is good, create the release branch and commit the generated pinned `versions.json` to the repository
  3. Rebuilds of the project using the pinned version are now possible by setting the :cmake:variable:`RAPIDS_CMAKE_CPM_OVERRIDE_VERSION_FILE` to the path of the generated pinned versions file.

> rapids-cpm command line controls
> ######################################

Reviewer: ```suggestion
rapids-cpm command line controls
###########################
```

rapids-cpm offers multiple command line options to control behavior.

Some builds are performed fully offline and the default package and override urls
can't be used. In those cases you can use the variable :cmake:variable:`RAPIDS_CMAKE_CPM_OVERRIDE_VERSION_FILE` to provide a new versions.json that will be
used instead of ALL overrides specified in that project. This would allow you to
specify custom internal urls for all dependencies without modifying the project source code.

```
cmake -DRAPIDS_CMAKE_CPM_OVERRIDE_VERSION_FILE=<abs/path/to/custom/override_versions.json> ....
```

To request the generation of a pinned package override file without having to modify
the project use the :cmake:variable:`RAPIDS_CMAKE_CPM_PINNED_VERSIONS_FILE` variable:

```
cmake -DRAPIDS_CMAKE_CPM_PINNED_VERSIONS_FILE
```

> additional CPM command line controls

Reviewer: ```suggestion
Additional CPM command line controls
```

************************************

In addition any of the CPM options can be used with rapids-cpm. A

> full list of CPM options can be found in the `CPM README.md <https://github.com/cpm-cmake/CPM.cmake/blob/master/README.md>`_.

Reviewer: ```suggestion
full list of CPM options can be found in the `CPM README.md <https://github.com/cpm-cmake/CPM.cmake/blob/master/README.md>`_;
we document some of the most important ones below.
```
